src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,140 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+    private int capacity;
+
+    public ArrayList() {
+        elementData = ((T[]) new Object[DEFAULT_CAPACITY]);
+        capacity = DEFAULT_CAPACITY;
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        rangeCheckForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            checkCapacity();
+            addWithShift(value, index);
+        }
+    }
 
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't reach index out of capacity");
+        }
+    }
+
+    private void rangeCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't reach index out of capacity");
+        }
+    }
+
+    private void addWithShift(T value, int index) {
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+    }
+
+    private void checkCapacity() {
+        if (size == capacity) {
+            expandCapacity();
+        }
+    }
 
+    private void expandCapacity() {
+        int newCapacity = newCapacity();
+        T[] newElementData = ((T[]) new Object[newCapacity]);
+        System.arraycopy(elementData, 0, newElementData, 0, size);
+        elementData = newElementData;
+        capacity = newCapacity;
+    }
+
+    private int newCapacity() {
```

Comment:
```suggestion
    private int getNewCapacity() {
```

methods names should start from verbs